File: android/app/src/main/java/com/ooplab/exercises_fitfuel/LandmarkSmoother.kt
package com.ooplab.exercises_fitfuel

import com.google.mediapipe.tasks.components.containers.NormalizedLandmark

data class LandmarkPoint(val x: Float, val y: Float, val estimated: Boolean = false)

class LandmarkSmoother(
    landmarkCount: Int = 33,
    minCutoff: Float = 1.5f,
    beta: Float = 0.01f
) {
    private val xFilters = Array(landmarkCount) { OneEuroFilter(minCutoff, beta) }
    private val yFilters = Array(landmarkCount) { OneEuroFilter(minCutoff, beta) }

    fun smooth(landmarks: List<NormalizedLandmark>, timestampSec: Double): List<LandmarkPoint> =
        landmarks.mapIndexed { i, lm ->
            LandmarkPoint(
                x = xFilters[i].filter(lm.x(), timestampSec),
                y = yFilters[i].filter(lm.y(), timestampSec)
            )
        }

    fun reset() {
        xFilters.forEach { it.reset() }
        yFilters.forEach { it.reset() }
    }
}

File: ios/Runner/PostureEngine/LandmarkSmoother.swift
import Foundation

/// Per-landmark One Euro smoothing of the x/y channels. Direct port of the Kotlin
/// `LandmarkSmoother`, but takes `[RawLandmark]` instead of MediaPipe types.
final class LandmarkSmoother {
    private let xFilters: [OneEuroFilter]
    private let yFilters: [OneEuroFilter]

    init(landmarkCount: Int = 33, minCutoff: Float = 1.5, beta: Float = 0.01) {
        xFilters = (0..<landmarkCount).map { _ in OneEuroFilter(minCutoff: minCutoff, beta: beta) }
        yFilters = (0..<landmarkCount).map { _ in OneEuroFilter(minCutoff: minCutoff, beta: beta) }
    }

    func smooth(_ landmarks: [RawLandmark], timestampSec: Double) -> [LandmarkPoint] {
        landmarks.enumerated().map { i, lm in
            LandmarkPoint(
                x: xFilters[i].filter(lm.x, timestampSec: timestampSec),
                y: yFilters[i].filter(lm.y, timestampSec: timestampSec)
            )
        }
    }

    func reset() {
        xFilters.forEach { $0.reset() }
        yFilters.forEach { $0.reset() }
    }
}
